### PR TITLE
Permit creation of LUKS devices

### PR DESCRIPTION
**Pull request by mjg59**
Add schema and support code for creating and activating LUKS devices on first boot

### CODE REVIEW

File: config/types/luks.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
+// Copyright 2016 CoreOS, Inc.
```
Comment (crawford) [inline config/types/luks.go]: 2017

File: internal/vendor.manifest [hunk 2]
```diff
@@ -4,6 +4,7 @@ github.com/coreos/go-systemd/dbus                      9b2f329b69ae0292a30f42686
 github.com/coreos/go-systemd/unit                      9b2f329b69ae0292a30f426867494f38af617a42
 github.com/coreos/update-ssh-keys/authorized_keys_d    539317dc5100bc599dab125b1e644a2b017b2b35
 github.com/godbus/dbus                                 41608027bdce7bfa8959d653a00b954591220e67
+github.com/martinjungblut/cryptsetup                   8c360554f611d89bf0fd81519d5a35d4de68c414
```
Comment (crawford) [inline internal/vendor.manifest]: Can you split this into a separate commit?

File: config/types/luks.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+var (
+	ErrLuksNoDevice     = errors.New("no device provided for encrypted volume setup")
+	ErrLuksKeyAndSource = errors.New("luks volume has both key and keysource defined")
+)
+
+type Luks struct {
+	Name      string `json:"name,omitempty"`
+	Device    Path   `json:"device,omitempty"`
+	Key       string `json:"key,omitempty"`
```
Comment (crawford) [inline config/types/luks.go]: These two should probably be collapsed into one key. Maybe use a URL and have immediate keys represented as a Data URL.

File: config/types/luks.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	KeySource string `json:"keysource,omitempty"`
+}
+
+func (l Luks) Validate() report.Report {
+	if len(l.Device) == 0 {
+		return report.ReportFromError(ErrLuksNoDevice, report.EntryError)
+	}
+	if l.Key != "" && l.KeySource != "" {
```
Comment (crawford) [inline config/types/luks.go]: We should probably also check for the case where there is no key.

File: internal/exec/stages/disks/disks.go [hunk 5]
```diff
@@ -222,6 +230,107 @@ func (s stage) createRaids(config types.Config) error {
...
+
+	err, device := cryptsetup.Init(string(luks.Device))
+	if err != nil {
+		return fmt.Errorf("Unable to initialise cryptsetup: %v", err)
+	}
+	params := cryptsetup.LUKSParams{Hash: "sha256",
+		Data_alignment: 0,
+		Data_device:    ""}
+
+	err = device.FormatLUKS("aes", "xts-plain64", "", "", 256/8, params)
```
Comment (crawford) [inline internal/exec/stages/disks/disks.go]: I would be nice to expose this in the config. We can add that later though.

**Comment by mjg59** [comment]
Done

File: internal/exec/stages/disks/disks.go [hunk 6]
```diff
@@ -222,6 +231,82 @@ func (s stage) createRaids(config types.Config) error {
 	return nil
 }
 
+// createLukss creates any LUKS partitions described in config.Storage.Luks
+func (s stage) createLukss(config types.Config) error {
+	lukss := make([]types.Luks, 0, len(config.Storage.Luks))
+	for _, luks := range config.Storage.Luks {
+		if len(luks.Device) > 0 {
```
Comment (vcaputo) [inline internal/exec/stages/disks/disks.go]: first, this is cool, adding luks support is great.

I was just drive-by glancing at this and you seem to be redundantly validating the device length here, why?  let the validation worry about that, you should probably just use the device paths.

and move the `if len(lukss) == 0 { return nil }` from down below to up top and instead checking `len(config.Storage.Luks)`.
